core/commonMain/src/kotlinx/serialization/internal/ElementMarker.kt

```diff
@@ -0,0 +1,93 @@
+package kotlinx.serialization.internal
+
+import kotlinx.serialization.ExperimentalSerializationApi
+import kotlinx.serialization.InternalSerializationApi
+import kotlinx.serialization.descriptors.SerialDescriptor
+import kotlinx.serialization.encoding.CompositeDecoder
+
+@InternalSerializationApi
+@OptIn(ExperimentalSerializationApi::class)
+public abstract class ElementMarker(private val descriptor: SerialDescriptor) {
+    /*
+    Element decoding marks from given bytes.
+    The element number is the same as the bit position.
+    Marks for the lowest 64 elements are always stored in a single Long value, higher elements stores in long array.
+     */
+    private var lowerMarks: Long
+    private val highMarksArray: LongArray?
+
+    init {
+        val elementsCount = descriptor.elementsCount
+        if (elementsCount <= Long.SIZE_BITS) {
+            lowerMarks = if (elementsCount == Long.SIZE_BITS) {
+                // number of bits in the mark is equal to the number of fields
+                0L
+            } else {
+                // (1 - elementsCount) bits are always 1 since there are no fields for them
+                -1L shl elementsCount
+            }
+            highMarksArray = null
+        } else {
+            lowerMarks = 0L
+            // (elementsCount - 1) because only one Long value is needed to store 64 fields etc
+            val slotsCount = (elementsCount - 1) / Long.SIZE_BITS
+            val elementsInLastSlot = elementsCount % Long.SIZE_BITS
+            val highMarks = LongArray(slotsCount)
+            // (elementsCount % Long.SIZE_BITS) == 0 this means that the fields occupy all bits in mark
+            if (elementsInLastSlot != 0) {
+                // all marks except the higher are always 0
+                highMarks[highMarks.lastIndex] = -1L shl elementsCount
+            }
+            highMarksArray = highMarks
+        }
+    }
+
+    protected abstract fun isPoppedElement(descriptor: SerialDescriptor, index: Int): Boolean
+
+    public fun popUnmarkedIndex(): Int {
```

Comment:
WDYT of naming like `takeUnmarkedIndex`/`nextUnmarkedIndex`/`pollUnmarkedElement` and corresponding, probably `elementNeedsMarking`. 'pop' strongly associates with a stack, but this structure is more like a queue

Reply:
I originally named the function `nextUnmarkedIndex`, but since it not only returns the index, but also marks the element as read, I decided to give it a more actual name. The variants were `nextElementIndexAndMark` or `popUnmarkedIndex` - decided to leave the second.

Reply:
I agree with Leonya, `nextUnmarkedIndex` sounds more familiar and iterator-ish rather than stack-like `pop`